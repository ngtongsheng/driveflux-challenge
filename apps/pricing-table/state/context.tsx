import React, { createContext, useReducer, Dispatch } from 'react';
import { Pricing } from '@driveflux-code-challenge/interfaces';

export const START_EDIT = 'START_EDIT';
export const CANCEL_EDIT = 'CANCEL_EDIT';
export const UPDATE_PRICING_FIELD = 'UPDATE_PRICING_FIELD';
export const RESET_ALL_PRICINGS = 'RESET_ALL_PRICINGS';
export const START_SAVE_PRICINGS = 'START_SAVE_PRICINGS';
export const SUCCESS_GET_PRICINGS = 'SUCCESS_GET_PRICINGS';
export const SUCCESS_SAVE_PRICINGS = 'SUCCESS_SAVE_PRICINGS';

interface Action {
  type: string;
  payload?: PricingField | Pricings;
}

interface PricingField {
  id: string;
  field: string;
  value: string;
}

interface Pricings {
  snapshot?: Pricing[];
  pricings: Pricing[];
}

interface PricingState extends Pricings {
  readOnly: boolean;
  isSaving: boolean;
}

const initialState: PricingState = {
  snapshot: [],
  pricings: [],
  readOnly: true,
  isSaving: false
};

const reducer = (state: PricingState, { type, payload }: Action): PricingState => {
  switch (type) {
    case SUCCESS_GET_PRICINGS: {
      const { pricings } = payload as Pricings;
      return {
        ...state,
        pricings
      };
    }

    case START_EDIT: {
      const { pricings } = state;
      return {
        ...state,
        snapshot: [...pricings],
        readOnly: false
      };
    }

    case CANCEL_EDIT: {
      const { snapshot } = state;
      return {
        ...state,
        snapshot: [],
        pricings: [...snapshot],
        readOnly: true
      };
    }

    case UPDATE_PRICING_FIELD: {
      const { id, field, value } = payload as PricingField;
      const newValue = Math.max(0, parseFloat(value));

      const pricings = state.pricings.map(pricing => {
        if (pricing.id !== id) {
          return pricing;
        }

        if (field !== "lite") {
          return {
            ...pricing,
            [field]: String(newValue)
          }
        }

        return {
          ...pricing,
          [field]: String(newValue),
          standard: String(newValue * 2),
          unlimited: String(newValue * 3),
        }
      })

      return {
        ...state,
        pricings
      };
    }

    case RESET_ALL_PRICINGS: {
      const pricings = state.pricings.map(pricing => ({
        ...pricing,
        lite: "0",
        standard: "0",
        unlimited: "0"
      }))

      return {
        ...state,
        pricings
      };
    }

    case START_SAVE_PRICINGS: {
      return {
        ...state,
        isSaving: true
      };
    }

    case SUCCESS_SAVE_PRICINGS: {
      return {
        ...state,
        isSaving: false,
        readOnly: true,
        snapshot: []
      };
    }

    default:
      return state;
  }
};

export const PricingContext = createContext<[s: PricingState, d: Dispatch<Action>]>(
  [null, null]
);

export const PricingContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <PricingContext.Provider value={[state, dispatch]}>
      {children}
    </PricingContext.Provider>
  );
};
