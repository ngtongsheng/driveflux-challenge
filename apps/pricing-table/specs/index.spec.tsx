import React from 'react';
import { render } from '@testing-library/react';

import Index from '../pages/index';
import { PricingContextProvider } from '../state/context';

describe('Index', () => {
  it('should render successfully', () => {
    const { baseElement } = render(
      <PricingContextProvider
        overrideState={{
          pricings: [
            {
              id: 'test',
              label: 'test',
              lite: '0',
              standard: '0',
              unlimited: '0',
            },
          ],
        }}
      >
        <Index />
      </PricingContextProvider>
    );

    expect(baseElement).toBeTruthy();
  });
});
