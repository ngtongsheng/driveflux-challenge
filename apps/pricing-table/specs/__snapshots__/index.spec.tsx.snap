// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index should render successfully 1`] = `
<body>
  <div>
    <div
      class="columns is-vcentered"
    >
      <div
        class="column"
      >
        <div
          class="title is-4 is-uppercase"
        >
          Pricing
        </div>
      </div>
      <div
        class="column is-narrow"
      >
        <div
          class="buttons"
        >
          <button
            class="jsx-2282515019 button is-dark"
          >
            Edit
          </button>
        </div>
      </div>
    </div>
    <table
      class="jsx-1459946533 table is-narrow is-hoverable is-fullwidth"
    >
      <thead
        class="jsx-1459946533"
      >
        <tr
          class="is-uppercase"
        >
          <th>
            Plan
          </th>
          <th>
            Lite
          </th>
          <th>
            Standard
          </th>
          <th>
            Unlimited
          </th>
        </tr>
      </thead>
      <tbody
        class="jsx-1459946533"
      >
        <tr>
          <td>
            test
          </td>
          <td>
            <input
              class="jsx-3324638787 input"
              name="lite"
              readonly=""
              type="number"
              value="0"
            />
          </td>
          <td>
            <input
              class="jsx-3324638787 input"
              name="standard"
              readonly=""
              type="number"
              value="0"
            />
          </td>
          <td>
            <input
              class="jsx-3324638787 input"
              name="unlimited"
              readonly=""
              type="number"
              value="0"
            />
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</body>
`;
